File: secure-code-review/vulnerable-examples/javascript/auth/jwt_validation.js
/**
 * Insecure JWT Validation in Express — vulnerable and secure variants.
 *
 * Scenario: an Express middleware that extracts the current user from
 * the `Authorization: Bearer <jwt>` header. Every protected route
 * depends on it. Errors here are full-compromise bugs.
 *
 * Failure modes demonstrated:
 *
 *   1. jwt.decode() used as a validation step. It does not validate
 *      — it only decodes. Any attacker-controlled payload is accepted
 *      as truth.
 *
 *   2. jwt.verify() without an explicit `algorithms` option. In
 *      older jsonwebtoken versions (and still reachable in some
 *      misconfigurations), this enables the classic
 *      RS256-as-HS256 algorithm-confusion attack: the attacker
 *      signs the token with the server's PUBLIC key using HS256,
 *      and the server — configured with the public key as the HMAC
 *      secret — accepts it.
 *
 *   3. No expiry check, so a 90-day-old stolen token is still valid.
 *
 *   4. `kid` header trusted blindly — the server fetches a key from
 *      whatever URL the `kid` points at, creating an arbitrary-key
 *      validation oracle.
 *
 * The secure version uses RS256 with JWKS lookup constrained to the
 * authorization server, pins algorithm and issuer and audience,
 * enforces expiry, tolerates small clock skew, and returns a sanitized
 * error to the client while logging the real reason server-side.
 */

const jwt = require("jsonwebtoken");
const jwksClient = require("jwks-rsa");

const EXPECTED_ISSUER = "https://auth.example.com/";
const EXPECTED_AUDIENCE = "api.example.com";

// ===========================================================================
// VULNERABLE middleware
// ===========================================================================

function authVulnerable(req, res, next) {
  const header = req.headers.authorization || "";
  const token = header.replace(/^Bearer /, "");

  // BUG 1: jwt.decode does NOT verify the signature. Any attacker-
  // constructed token with a valid shape is accepted.
  const payload = jwt.decode(token);
  if (!payload) return res.status(401).json({ error: "missing token" });

  // BUG 2: `role` is trusted verbatim from the token with no
  // validation that the token was signed by the authorization server.
  req.user = {
    id: payload.sub,
    role: payload.role,
    tenantId: payload.tenant_id,
  };
  next();
}

// Equally common variation — jwt.verify with no algorithms option.
//
//   jwt.verify(token, SHARED_SECRET, (err, decoded) => { ... });
//
// Against a library version that defaults to permissive algorithms,
// an attacker can craft an HS256-signed token using the server's
// RSA public key and the server validates it because it thinks the
// "secret" IS the public key.

// ===========================================================================
// SECURE middleware — layered defenses against each failure mode.
// ===========================================================================

// JWKS client constrained to the authorization server's jwks_uri.
// Caching reduces the chance of upstream dependence during a spike;
// rate limiting prevents an attacker from exhausting the cache with
// random `kid` values.
const keyClient = jwksClient({
  jwksUri: `${EXPECTED_ISSUER}.well-known/jwks.json`,
  cache: true,
  cacheMaxAge: 10 * 60 * 1000,
  rateLimit: true,
  jwksRequestsPerMinute: 10,
  timeout: 3000,
});

function getSigningKey(header, cb) {
  // `kid` must be a non-empty string; refuse anything else.
  if (!header.kid || typeof header.kid !== "string") {
    return cb(new Error("missing kid"));
  }
  keyClient.getSigningKey(header.kid, (err, key) => {
    if (err) return cb(err);
    cb(null, key.getPublicKey());
  });
}

function authSecure(req, res, next) {
  const header = req.headers.authorization || "";
  const match = /^Bearer (.+)$/.exec(header);
  if (!match) {
    return res.status(401).json({ error: "missing token" });
  }
  const token = match[1];

  jwt.verify(
    token,
    getSigningKey,
    {
      algorithms: ["RS256"],          // explicit allowlist — kills
                                      // alg=none and RS256-as-HS256
      issuer: EXPECTED_ISSUER,        // rejects tokens from other IdPs
      audience: EXPECTED_AUDIENCE,    // rejects tokens for other services
      clockTolerance: 30,             // seconds — absorb normal drift,
                                      // not a replay window
      complete: false,
    },
    (err, decoded) => {
      if (err) {
        // Log the real error server-side for debugging and detection.
        req.log?.warn(
          { reqId: req.id, err: err.message },
          "jwt rejected"
        );
        // Return a generic message to the client — no leak of whether
        // expiry, signature, or structure failed.
        return res.status(401).json({ error: "invalid token" });
      }

      // Additional application-specific invariants.
      if (!decoded.sub) {
        return res.status(401).json({ error: "invalid token" });
      }

      // Minimal, whitelisted fields. `role` is trusted at this level
      // only for non-privileged decisions; high-impact authorization
      // re-fetches the current role from the identity service.
      req.user = {
        id: decoded.sub,
        role: decoded.role ?? "user",
        tenantId: decoded.tenant_id,
        issuedAt: decoded.iat,
        expiresAt: decoded.exp,
      };

      next();
    }
  );
}

// ===========================================================================
// A per-route scope/role check, since authentication alone is not
// authorization. The reviewer's rule: every protected handler
// explicitly states what role or scope is required — never inferred
// from "the user is logged in."
// ===========================================================================

function requireRole(role) {
  return (req, res, next) => {
    if (!req.user) return res.status(401).json({ error: "unauthenticated" });
    if (req.user.role !== role) {
      req.log?.info(
        { reqId: req.id, userId: req.user.id, required: role, actual: req.user.role },
        "forbidden"
      );
      return res.status(403).json({ error: "forbidden" });
    }
    next();
  };
}

// Usage:
//   app.use(authSecure);
//   app.post("/admin/users", requireRole("admin"), adminHandler);

// ===========================================================================
// Common patterns that look secure but are not, worth calling out in
// review:
//
// - `jwt.verify(token, getKeyFromKid, {...})` where `getKeyFromKid`
//   follows the URL embedded in the `kid` claim. This is an
//   arbitrary-key oracle — attacker controls which key you validate
//   against. Always pull keys from a fixed JWKS URI, not from the
//   token itself.
//
// - Trusting the `role` claim for high-privilege operations. The JWT
//   was issued minutes or hours ago; the user's entitlements may have
//   changed. For admin actions, re-check authorization against the
//   identity service at the moment of the action.
//
// - "We only use this behind the API gateway which already validated
//   the JWT." Until someone makes the service reachable from another
//   path — a cron job, a test harness, a migration script. Validate
//   on every entry point.
//
// - jwt.verify without `audience` and `issuer`. A token issued for a
//   different service in the same IdP passes signature validation.
//   You now accept another team's tokens.
// ===========================================================================

module.exports = { authVulnerable, authSecure, requireRole };

File: secure-code-review/vulnerable-examples/python/auth/jwt_validation.py
"""
Insecure JWT Validation — vulnerable and secure variants, side by side.

Scenario: a FastAPI dependency that extracts the current user from an
incoming `Authorization: Bearer <jwt>` header. This is the
authentication function every other handler depends on. Getting it
wrong is a full-compromise bug.

Two real-world failure modes are shown:

  1. `algorithms=["none"]` or missing algorithm pin — attacker replaces
     the signature with an empty string and sets `alg: none`, bypassing
     signature verification entirely. Still observed in the wild in
     2026; the PyJWT maintainers had to add active defenses because of
     how often it was misused.

  2. No expiry check. Token issued months ago is still valid.

  3. HS256 with a shared secret from an env var that was leaked in a
     prior git commit. The attacker forges tokens offline.

The secure version pins the algorithm, validates every claim, rejects
expired tokens, and uses an asymmetric algorithm (RS256) with a public
key so even a compromised application server cannot mint new tokens.
"""

import os
import time

import jwt  # PyJWT
from jwt import ExpiredSignatureError, InvalidTokenError


EXPECTED_ISSUER = "https://auth.example.com/"
EXPECTED_AUDIENCE = "api.example.com"


# ---------------------------------------------------------------------------
# VULNERABLE: accepts alg=none, no expiry check, trusts arbitrary claims.
# ---------------------------------------------------------------------------
def current_user_vulnerable(token: str) -> dict:
    """
    Exploit A — alg=none:

        Take any JWT with a known payload structure, base64-decode it,
        modify the payload to your target user, re-encode, then send as:

            <header_with_alg_none>.<forged_payload>.

        PyJWT accepts it because `algorithms` is None, so every alg is
        allowed — including "none", which skips signature verification.

    Exploit B — signature-skip option:

        options={"verify_signature": False} is here for a "debugging"
        reason a junior dev added during an incident and never removed.
        It means every signed or unsigned token is trusted.

    Exploit C — no expiry:

        A token issued 18 months ago during pen testing is still valid
        because `verify_exp` was never turned on. Any captured token
        becomes a perpetual credential.
    """
    payload = jwt.decode(
        token,
        # Note: no key passed, because verify_signature is off.
        options={"verify_signature": False, "verify_exp": False},
    )
    return {
        "user_id": payload["sub"],
        "role": payload.get("role", "user"),  # fully attacker-controlled
        "tenant_id": payload.get("tenant_id"),
    }


# ---------------------------------------------------------------------------
# SECURE: pinned algorithm, asymmetric key, explicit claim validation.
# ---------------------------------------------------------------------------
def _load_public_key() -> str:
    """
    In production: load from JWKS and cache per `kid`. For brevity this
    example reads from an env var populated at container start.
    """
    pem = os.environ.get("JWT_PUBLIC_KEY_PEM")
    if not pem:
        raise RuntimeError("JWT_PUBLIC_KEY_PEM is required")
    return pem


def current_user_secure(token: str) -> dict:
    """
    Properties of this function, each tied to a specific failure mode:

      - `algorithms=["RS256"]` — explicit allowlist. alg=none is
        rejected. Algorithm-confusion (RS256-as-HS256) is rejected
        because HS256 is not in the list.

      - `key=<public_key>` — asymmetric key. Even if this process is
        compromised, the attacker cannot mint new tokens without the
        private key, which lives in the authorization server.

      - `options={"require": [...]}` — required claims must be present.
        Missing `exp` is a token we reject, not a token we treat as
        eternal.

      - `verify_iss`, `verify_aud` — prevent a token issued for a
        different audience (e.g., the consumer app's token) from being
        accepted by this service.

      - `leeway=30` — 30-second clock skew tolerance. Without this you
        get intermittent failures during real clock drift; with a much
        larger value you give attackers a longer replay window.

      - Claim whitelist on the return — we build our user object from
        only the claims we validated. We do not echo arbitrary claims
        back to the rest of the application.
    """
    try:
        payload = jwt.decode(
            token,
            _load_public_key(),
            algorithms=["RS256"],
            audience=EXPECTED_AUDIENCE,
            issuer=EXPECTED_ISSUER,
            leeway=30,
            options={
                "require": ["exp", "iat", "iss", "aud", "sub"],
                "verify_signature": True,
                "verify_exp": True,
                "verify_iat": True,
                "verify_iss": True,
                "verify_aud": True,
            },
        )
    except ExpiredSignatureError:
        raise PermissionError("token expired") from None
    except InvalidTokenError as e:
        # Intentionally do not echo the underlying error to the client;
        # log it with a request ID and return a generic 401.
        raise PermissionError("invalid token") from e

    # Minimal, validated user record. `role` is fetched fresh from the
    # authorization service for high-privilege operations; the JWT
    # claim is only used for low-sensitivity routing.
    return {
        "user_id": payload["sub"],
        "role": payload.get("role", "user"),
        "tenant_id": payload.get("tenant_id"),
        "issued_at": payload["iat"],
        "expires_at": payload["exp"],
    }


# ---------------------------------------------------------------------------
# A lightweight test harness. Run this file to see the vulnerable
# version accept a forged token that the secure version rejects.
# ---------------------------------------------------------------------------
def _demo():
    from cryptography.hazmat.primitives.asymmetric import rsa
    from cryptography.hazmat.primitives import serialization

    # Generate a keypair just for the demo.
    private_key = rsa.generate_private_key(public_exponent=65537, key_size=2048)
    public_pem = private_key.public_key().public_bytes(
        encoding=serialization.Encoding.PEM,
        format=serialization.PublicFormat.SubjectPublicKeyInfo,
    ).decode()
    private_pem = private_key.private_bytes(
        encoding=serialization.Encoding.PEM,
        format=serialization.PrivateFormat.PKCS8,
        encryption_algorithm=serialization.NoEncryption(),
    ).decode()

    os.environ["JWT_PUBLIC_KEY_PEM"] = public_pem

    now = int(time.time())

    # A legitimately-signed token.
    good_token = jwt.encode(
        {
            "sub": "user-123", "role": "user", "tenant_id": "t-1",
            "iss": EXPECTED_ISSUER, "aud": EXPECTED_AUDIENCE,
            "iat": now, "exp": now + 900,
        },
        private_pem, algorithm="RS256",
    )

    # A forged "alg=none" token with admin role.
    import base64, json
    def b64(x):
        return base64.urlsafe_b64encode(
            json.dumps(x, separators=(",", ":")).encode()
        ).rstrip(b"=").decode()
    forged_token = (
        b64({"typ": "JWT", "alg": "none"}) + "."
        + b64({"sub": "attacker", "role": "admin", "tenant_id": "t-1",
               "iss": EXPECTED_ISSUER, "aud": EXPECTED_AUDIENCE,
               "iat": now - 3600, "exp": now - 60})
        + "."
    )

    # Vulnerable accepts the forgery AND the expired forged token.
    print("VULNERABLE (forged):", current_user_vulnerable(forged_token))
    # → {'user_id': 'attacker', 'role': 'admin', ...}

    # Secure rejects the forgery.
    try:
        current_user_secure(forged_token)
    except PermissionError as e:
        print("SECURE (forged):    rejected —", e)

    # Both accept the good token.
    print("VULNERABLE (good):  ", current_user_vulnerable(good_token))
    print("SECURE (good):      ", current_user_secure(good_token))


if __name__ == "__main__":
    _demo()
